File: examples/factors_abs_complex.py
#
# jython examples for jas.
# $Id$
#

import sys;

from java.lang import System

from jas import Ring, PolyRing
from jas import QQ, AN
from jas import terminate, startLog

# polynomial examples: absolute factorization over Q(i)

Qr = PolyRing(QQ(),"i",PolyRing.lex);
print "Qr    = " + str(Qr);
[e,a] = Qr.gens();
print "e     = " + str(e);
print "a     = " + str(a);
print;

imag = a**2 + 1;
print "imag  = " + str(imag);
Qi = AN(imag,field=True);
print "Qi    = " + str(Qi.factory());
[one,i] = Qi.gens();
print "one   = " + str(one);
print "i     = " + str(i);
print;

r = PolyRing(Qi,"x",PolyRing.lex)
print "r    = " + str(r);
#is automatic: [one,i,x] = r.gens();
print "one   = " + str(one);
print "i     = " + str(i);
print "x     = " + str(x);
print;


#f = x**7 - 1;
#f = x**6 + x**5 + x**4 + x**3 + x**2 + x + 1;
#f = x**2 - i;
#f = x**3 - i;
#f = x**2 + 1;
#f = x**5 - 1;
#f = x**4 + x**3 + x**2 + x + (1,);
#f = ( x**2 - i - 1 ) * ( x**2 + i + 1 );
f = ( x**2 - i ) * ( x**2 + i + 1 );

print "f = ", f;
print;

#startLog();

t = System.currentTimeMillis();
G = r.factors(f);
t = System.currentTimeMillis() - t;
#print "G = ", G;
#print "factor time =", t, "milliseconds";

f2 = one;
for h, i in G.iteritems():
    print "h**i = (", h, ")**" + str(i);
    h = h**i;
    f2 = f2*h;
#print "f2 = ", f2;
print;

if cmp(f,f2) == 0:
    print "factor time =", t, "milliseconds,", "isFactors(f,g): true" ;
else:
    print "factor time =", t, "milliseconds,", "isFactors(f,g): ",  cmp(f,f2);
print;


startLog();

t = System.currentTimeMillis();
G = r.factorsAbsolute(f);
t = System.currentTimeMillis() - t;
print "G = ", G.toScript();
print
print "factor time =", t, "milliseconds";
print

#sys.exit();
#startLog();

terminate();

File: examples/factors_abs_complex.rb
#
# jruby examples for jas.
# $Id$
#

require "examples/jas"

# polynomial examples: absolute factorization over Q(i)

Qr = PolyRing.new(QQ(),"i",PolyRing.lex);
puts "Qr    = " + str(Qr);
e,a = Qr.gens();
puts "e     = " + str(e);
puts "a     = " + str(a);
imag = a**2 + 1;
puts "imag  = " + str(imag);
Qi = AN(imag,true);
puts "Qi    = " + str(Qi.factory());
one,i = Qi.gens();
puts "one   = " + str(one);
puts "i     = " + str(i);
puts;


r = PolyRing.new(Qi,"x",PolyRing.lex)
puts "r    = " + str(r);

one,i,x = r.gens();
puts "one   = " + str(one);
puts "i     = " + str(i);
puts "x     = " + str(x);
puts;


#f = x**7 - 1;
#f = x**6 + x**5 + x**4 + x**3 + x**2 + x + 1;
#f = x**2 - i;
#f = x**3 - i;
#f = x**2 + 1;
#f = x**5 - 1;
#f = x**4 + x**3 + x**2 + x + (1,);
#f = ( x**2 - i - 1 ) * ( x**2 + i + 1 );
f = ( x**2 - i ) * ( x**2 + i + 1 );

puts "f = " + str(f);
puts;

#startLog();

t = System.currentTimeMillis();
g = r.factors(f);
t = System.currentTimeMillis() - t;
#puts "G = " + str(g);
#puts "factor time = " + str(t) + " milliseconds";

f2 = one;
for h, i in g
    puts "h**i = ("  + str(h) + ")**" + str(i);
    h = h**i;
    f2 = f2*h;
end
#puts "f2 = " + str(f2);
puts;

puts "factor time = " + str(t) + " milliseconds, " + "isFactors(f,g): " + str( f==f2 );
puts;


startLog();

t = System.currentTimeMillis();
g = r.factorsAbsolute(f);
t = System.currentTimeMillis() - t;
puts "G = ", g.toScript();
puts
puts "factor time = " + str(t) + " milliseconds";
puts

#sys.exit();
#startLog();

terminate();
